src/survey/ia-feedback-survey.ts

```diff
@@ -0,0 +1,843 @@
+                @click=${this.cancel}
+              >
+                ${msg('Cancel')}
+              </button>
+              <input
+                type="submit"
+                id="submit-button"
+                class="cta-button"
+                tabindex="0"
+                ?disabled=${shouldDisableControls}
+                .value=${submitButtonText}
+              />
+            </div>
+          </form>
+        </div>
+        <div
+          class="focus-trap"
+          tabindex="0"
+          @focus=${this.focusFirstFormElement}
+        ></div>
+      </div>
+    `;
+  }
+
+  /**
+   * Assigns focus to the first slotted element in the form.
+   */
+  private focusFirstFormElement(): void {
+    this.assignedElements[0]?.focus();
+  }
+
+  /**
+   * Assigns focus to the submit button at the end of the form.
+   */
+  private focusSubmitButton(): void {
+    this.submitButton?.focus();
+  }
+
+  /**
+   * Handler for when the background layer outside the popup is clicked, treated
+   * as though the Cancel button were clicked.
+   */
+  private backgroundClicked(e: MouseEvent) {
+    if (!(e.target instanceof Node)) return;
+    if (this.popup?.contains(e.target)) return;
+    this.cancel(e);
+  }
+
+  /**
+   * Closes the survey popup and resets any submission error state.
+   */
+  private cancel(e: Event) {
+    e.preventDefault();
+    this.closePopup();
+    if (this.submissionState !== 'submitted') this.resetSubmissionState();
+  }
+
+  /**
+   * Returns a boolean indicating whether all required survey questions have received a
+   * valid response.
+   *
+   * - Questions with `required: true` and `type: 'vote'` must have an up/down vote recorded.
+   * - Questions with `required: true` and `type: 'comment'` must have a comment recorded.
+   * - Questions without `required: true`, or having `type: 'extra'`, are always valid.
+   */
+  private validate(): boolean {
+    return this.assignedElements
+      .filter(canBeValidated)
+      .map(elmt => elmt.validate())
+      .every(v => v);
+  }
+
+  /**
+   * Handles survey form submit events by attempting to submit the responses
+   * to the server, provided all required responses have been entered.
+   */
+  private async submit(e?: Event) {
+    e?.preventDefault();
+
+    if (!this.validate()) {
+      this.error = html`${IAFeedbackSurvey.ERROR_MESSAGE_MISSING_REQUIRED_INPUT}`;
+      this.setSubmissionState('error');
+      return;
+    }
+
+    const { surveyIdentifier, submitTimeout, featureFeedbackService } = this;
+    this.error = undefined;
+
+    if (!surveyIdentifier) {
+      throw new Error('surveyIdentifier is required');
+    }
+
+    if (!featureFeedbackService) {
+      throw new Error('featureFeedbackService is required');
+    }
+
+    const recaptchaWidgetPromise = this.getRecaptchaWidget();
+    let recaptchaWidget: RecaptchaWidgetInterface | undefined;
+    try {
+      recaptchaWidget = await timedPromise(
```

Comment:
Out of curiosity, were you seeing timeouts of the recaptcha loading or it taking too long or something?

Reply:
What I noticed initially was that the demo app's recaptcha requests always (predictably) fail internally when using a dummy site key, but those failures don't necessarily cause the recaptcha widget promise to be rejected. I'm not certain _why_ that's the case, but it seemed broadly sensible to apply a time limit to the operation regardless, so that we can at least render an error message if things take too long (rather than just spinning in the Submitting... state forever).

Reply:
Ahh interesting! I wonder if at some point we might want to build this functionality into the recaptchaManager since it's likely going to be a common issue.